Lock Ownership Model

# Use Case

I'm trying to do is use the `LockManager`, to create a lock, and pass it to a tokio channel. 

Something like

```rust
let entry: T;
let lockManager = LockManager::new();
let lock = lockManager.lock();
channel.send( (entry, lock) );
```

# Issue

However rust is complaining that the `lockManager` will go out of scope. So I tried to send both the `lockManager` and the lock through the channel.

Changing the logic to

```rust
let entry: T;
let lockManager = Arc::new(LockManager::new());
let lock: Lock<'_> = lockManager.lock(..., ...);
channel.send( (entry, lock) );
```
does not work.

# Proposed Change

I think the developer experience could be improved if you did something like this:
```rust
pub struct Lock<'a> {
    /// The resource to lock. Will be used as the key in Redis.
    pub resource: Vec<u8>,
    /// The value for this lock.
    pub val: Vec<u8>,
    /// Time the lock is still valid.
    /// Should only be slightly smaller than the requested TTL.
    pub validity_time: usize,
    /// Used to limit the lifetime of a lock to its lock manager.
    pub lock_manager: Arc<LockManagerInner>,
}

pub struct LockManagerInner { 
...
}

pub struct LockManager { 
   pub lock_manager: Arc<LockManagerInner>
}
```
This would allow a channel responsible for creating locks, a channel for doing other stuff, and a channel for closing locks to be represented clearly without having to worry about lifetimes. 

This is the pattern that the [mongo](https://github.com/mongodb/mongo-rust-driver/blob/main/src/db.rs) driver for rust uses & it moves the complexity from the user to the crate.

Would you welcome a PR for this change?

I'm ingesting data from a Kafka topic, represented as a channel in Rust and need to acquire a lock on the given value, before modifying the value in redis/db, while still allowing other values to be processed concurrently.

In my production code, the function signature is a bit convoluted:
```rust
async fn stream<T: DeserializeOwned + Send + Sync + 'static + Debug + Clone, ...>(mut receiver: Receiver<T>, ...) {}
```
When I looked at your example, I thought the issue might have been the `static` lifetime requirement for T, but my simpler example is still throwing lifetime errors:

```rust
#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
struct ExampleStruct {
//Random data
}

async fn stream(mut receiver: Receiver<ExampleStruct>, lock_manager: LockManager) {
       //channel will be consumed by another spawned tokio thread, omitted for clarity
       let (tx, mut rx) = mpsc::channel(32);
        loop {
            match receiver.recv().await {
                Some(e) => {
                    let lock_managed = lock_manager.clone();
                    let sending = tx.clone();
                    
                    tokio::spawn(async move {
                        let l = lock_managed.lock(b"key", 10).await;
                        sending.send( (e, l) );
                    });

                }
                None => {}
            }
        }
    }
   
```

This is similar to what I'm going for and this is the error I get:
```
error[E0597]: `lock_managed` does not live long enough
let lock_managed = lock_manager.clone();
    |                         ------------ binding `lock_managed` declared here
    |                     let sending = tx.clone();
    |                         ------- lifetime `'1` appears in the type of `sending`
    |                     tokio::spawn(async move {
    |                         let l = lock_managed.lock(b"key", 10).await;
    |                                 ^^^^^^^^^^^^ borrowed value does not live long enough
    |                         sending.send( (e, l));
    |                         ---------------------- argument requires that `lock_managed` is borrowed for `'1`
    |                     });
    |                     - `lock_managed` dropped here while still borrowed
```

--- a/src/lock.rs
+++ b/src/lock.rs
@@ -732,4 +732,27 @@ mod tests {
             Err(_) => (), // Test passes
         }
     }
+
+    #[tokio::test]
+    async fn test_lock_send_lock_manager() {
+        let (_containers, addresses) = create_clients();
+        let rl = LockManager::new(addresses.clone());
+
+        let lock = rl
+            .lock(b"resource", std::time::Duration::from_millis(1000))
+            .await
+            .unwrap();
+
+        // Send the lock and entry through the channel
+        let (tx, mut rx) = tokio::sync::mpsc::channel(32);
+        tx.send(("some info", lock, rl)).await.unwrap();
+
+        let j = tokio::spawn(async move {
+            // Retrieve from channel and use
+            if let Some((_entry, lock, rl)) = rx.recv().await {
+                rl.unlock(&lock).await;
+            }
+        });
+        let _ = j.await;
+    }
 }

--- a/src/lock.rs
+++ b/src/lock.rs
@@ -66,7 +66,7 @@ struct LockManagerInner {
     quorum: u32,
 }
 
-#[derive(Debug, Clone)]
+#[derive(Debug)]
 pub struct Lock {
     /// The resource to lock. Will be used as the key in Redis.
     pub resource: Vec<u8>,
@@ -88,7 +88,7 @@ pub struct Lock {
 /// meaning that dropping the `LockGuard` will be a no-op.
 /// Under this circumstance, `LockManager::unlock` can be called manually using the inner `lock` at the appropriate
 /// point to release the lock taken in `Redis`.
-#[derive(Debug, Clone)]
+#[derive(Debug)]
 pub struct LockGuard {
     pub lock: Lock,
 }

--- a/src/lock.rs
+++ b/src/lock.rs
@@ -1,4 +1,5 @@
 use std::io;
+use std::sync::Arc;
 use std::time::{Duration, Instant};
 
 use futures::future::join_all;
@@ -53,15 +54,20 @@ pub enum LockError {
 /// and handles the Redis connections.
 #[derive(Debug, Clone)]
 pub struct LockManager {
-    /// List of all Redis clients
-    pub servers: Vec<Client>,
-    quorum: u32,
+    lock_manager_inner: Arc<LockManagerInner>,
     retry_count: u32,
     retry_delay: Duration,
 }
 
 #[derive(Debug, Clone)]
-pub struct Lock<'a> {
+struct LockManagerInner {
+    /// List of all Redis clients
+    pub servers: Vec<Client>,
+    quorum: u32,
+}
+
+#[derive(Debug, Clone)]
+pub struct Lock {
     /// The resource to lock. Will be used as the key in Redis.
     pub resource: Vec<u8>,
     /// The value for this lock.
@@ -70,7 +76,7 @@ pub struct Lock<'a> {
     /// Should only be slightly smaller than the requested TTL.
     pub validity_time: usize,
     /// Used to limit the lifetime of a lock to its lock manager.
-    pub lock_manager: &'a LockManager,
+    pub lock_manager: LockManager,
 }
 
 /// Upon dropping the guard, `LockManager::unlock` will be ran synchronously on the executor.
@@ -83,15 +89,15 @@ pub struct Lock<'a> {
 /// Under this circumstance, `LockManager::unlock` can be called manually using the inner `lock` at the appropriate
 /// point to release the lock taken in `Redis`.
 #[derive(Debug, Clone)]
-pub struct LockGuard<'a> {
-    pub lock: Lock<'a>,
+pub struct LockGuard {
+    pub lock: Lock,
 }
 
 /// Dropping this guard inside the context of a tokio runtime if `tokio-comp` is enabled
 /// will block the tokio runtime.
 /// Because of this, the guard is not compiled if `tokio-comp` is enabled.
 #[cfg(not(feature = "tokio-comp"))]
-impl Drop for LockGuard<'_> {
+impl Drop for LockGuard {
     fn drop(&mut self) {
         futures::executor::block_on(self.lock.lock_manager.unlock(&self.lock));
     }
@@ -111,8 +117,7 @@ impl LockManager {
             .collect();
 
         LockManager {
-            servers,
-            quorum,
+            lock_manager_inner: Arc::new(LockManagerInner { servers, quorum }),
             retry_count: DEFAULT_RETRY_COUNT,
             retry_delay: DEFAULT_RETRY_DELAY,
         }
@@ -204,14 +209,14 @@ impl LockManager {
         value: &[u8],
         ttl: usize,
         lock: T,
-    ) -> Result<Lock<'a>, LockError>
+    ) -> Result<Lock, LockError>
     where
         T: Fn(&'a Client) -> Fut,
         Fut: Future<Output = bool>,
     {
         for _ in 0..self.retry_count {
             let start_time = Instant::now();
-            let n = join_all(self.servers.iter().map(&lock))
+            let n = join_all(self.lock_manager_inner.servers.iter().map(&lock))
                 .await
                 .into_iter()
                 .fold(0, |count, locked| if locked { count + 1 } else { count });
@@ -228,16 +233,17 @@ impl LockManager {
                 - elapsed.as_secs() as usize * 1000
                 - elapsed.subsec_nanos() as usize / 1_000_000;
 
-            if n >= self.quorum && validity_time > 0 {
+            if n >= self.lock_manager_inner.quorum && validity_time > 0 {
                 return Ok(Lock {
-                    lock_manager: self,
+                    lock_manager: self.clone(),
                     resource: resource.to_vec(),
                     val: value.to_vec(),
                     validity_time,
                 });
             } else {
                 join_all(
-                    self.servers
+                    self.lock_manager_inner
+                        .servers
                         .iter()
                         .map(|client| Self::unlock_instance(client, resource, value)),
                 )
@@ -260,9 +266,10 @@ impl LockManager {
     ///
     /// Unlock is best effort. It will simply try to contact all instances
     /// and remove the key.
-    pub async fn unlock(&self, lock: &Lock<'_>) {
+    pub async fn unlock(&self, lock: &Lock) {
         join_all(
-            self.servers
+            self.lock_manager_inner
+                .servers
                 .iter()
                 .map(|client| Self::unlock_instance(client, &lock.resource, &lock.val)),
         )
@@ -278,7 +285,7 @@ impl LockManager {
     /// A user should retry after a short wait time.
     ///
     /// May return `LockError::TtlTooLarge` if `ttl` is too large.
-    pub async fn lock<'a>(&'a self, resource: &[u8], ttl: Duration) -> Result<Lock<'a>, LockError> {
+    pub async fn lock(&self, resource: &[u8], ttl: Duration) -> Result<Lock, LockError> {
         let val = self.get_unique_lock_id().map_err(LockError::Io)?;
         let ttl = ttl
             .as_millis()
@@ -297,11 +304,7 @@ impl LockManager {
     ///
     /// May return `LockError::TtlTooLarge` if `ttl` is too large.
     #[cfg(feature = "async-std-comp")]
-    pub async fn acquire<'a>(
-        &'a self,
-        resource: &[u8],
-        ttl: Duration,
-    ) -> Result<LockGuard<'a>, LockError> {
+    pub async fn acquire(&self, resource: &[u8], ttl: Duration) -> Result<LockGuard, LockError> {
         let lock = self.acquire_no_guard(resource, ttl).await?;
         Ok(LockGuard { lock })
     }
@@ -312,11 +315,11 @@ impl LockManager {
     /// or `LockManager::unlock` must be called to allow other clients to lock the same resource.
     ///
     /// May return `LockError::TtlTooLarge` if `ttl` is too large.
-    pub async fn acquire_no_guard<'a>(
-        &'a self,
+    pub async fn acquire_no_guard(
+        &self,
         resource: &[u8],
         ttl: Duration,
-    ) -> Result<Lock<'a>, LockError> {
+    ) -> Result<Lock, LockError> {
         loop {
             match self.lock(resource, ttl).await {
                 Ok(lock) => return Ok(lock),
@@ -327,11 +330,7 @@ impl LockManager {
     }
 
     /// Extend the given lock by given time in milliseconds
-    pub async fn extend<'a>(
-        &'a self,
-        lock: &Lock<'a>,
-        ttl: Duration,
-    ) -> Result<Lock<'a>, LockError> {
+    pub async fn extend(&self, lock: &Lock, ttl: Duration) -> Result<Lock, LockError> {
         let ttl = ttl
             .as_millis()
             .try_into()
@@ -413,8 +412,8 @@ mod tests {
 
         let rl = LockManager::new(addresses.clone());
 
-        assert_eq!(3, rl.servers.len());
-        assert_eq!(2, rl.quorum);
+        assert_eq!(3, rl.lock_manager_inner.servers.len());
+        assert_eq!(2, rl.lock_manager_inner.quorum);
     }
 
     #[tokio::test]
@@ -425,7 +424,7 @@ mod tests {
         let key = rl.get_unique_lock_id()?;
 
         let val = rl.get_unique_lock_id()?;
-        assert!(!LockManager::unlock_instance(&rl.servers[0], &key, &val).await);
+        assert!(!LockManager::unlock_instance(&rl.lock_manager_inner.servers[0], &key, &val).await);
 
         Ok(())
     }
@@ -438,10 +437,10 @@ mod tests {
         let key = rl.get_unique_lock_id()?;
 
         let val = rl.get_unique_lock_id()?;
-        let mut con = rl.servers[0].get_connection()?;
+        let mut con = rl.lock_manager_inner.servers[0].get_connection()?;
         redis::cmd("SET").arg(&*key).arg(&*val).execute(&mut con);
 
-        assert!(LockManager::unlock_instance(&rl.servers[0], &key, &val).await);
+        assert!(LockManager::unlock_instance(&rl.lock_manager_inner.servers[0], &key, &val).await);
 
         Ok(())
     }
@@ -454,10 +453,13 @@ mod tests {
         let key = rl.get_unique_lock_id()?;
 
         let val = rl.get_unique_lock_id()?;
-        let mut con = rl.servers[0].get_connection()?;
+        let mut con = rl.lock_manager_inner.servers[0].get_connection()?;
 
         redis::cmd("DEL").arg(&*key).execute(&mut con);
-        assert!(LockManager::lock_instance(&rl.servers[0], &*key, val.clone(), 1000).await);
+        assert!(
+            LockManager::lock_instance(&rl.lock_manager_inner.servers[0], &key, val.clone(), 1000)
+                .await
+        );
 
         Ok(())
     }
@@ -470,7 +472,7 @@ mod tests {
         let key = rl.get_unique_lock_id()?;
 
         let val = rl.get_unique_lock_id()?;
-        let mut con = rl.servers[0].get_connection()?;
+        let mut con = rl.lock_manager_inner.servers[0].get_connection()?;
         let _: () = redis::cmd("SET")
             .arg(&*key)
             .arg(&*val)
@@ -478,7 +480,7 @@ mod tests {
             .unwrap();
 
         let lock = Lock {
-            lock_manager: &rl,
+            lock_manager: rl.clone(),
             resource: key,
             val,
             validity_time: 0,

--- a/src/lock.rs
+++ b/src/lock.rs
@@ -1,5 +1,5 @@
 use std::io;
-use std::sync::Arc;
+use std::sync::{atomic::AtomicBool, atomic::Ordering, Arc};
 use std::time::{Duration, Instant};
 
 use futures::future::join_all;
@@ -77,6 +77,16 @@ pub struct Lock {
     pub validity_time: usize,
     /// Used to limit the lifetime of a lock to its lock manager.
     pub lock_manager: LockManager,
+    /// A flag indicating that this lock has been freed.
+    is_freed: AtomicBool,
+}
+
+impl Lock {
+    /// Retrieves the status of the lock, indicating whether it has been released.
+    /// **Note**: This status reflects only manual release by calling `unlock`.
+    pub fn is_freed(&self) -> bool {
+        self.is_freed.load(Ordering::Acquire)
+    }
 }
 
 /// Upon dropping the guard, `LockManager::unlock` will be ran synchronously on the executor.
@@ -239,6 +249,7 @@ impl LockManager {
                     resource: resource.to_vec(),
                     val: value.to_vec(),
                     validity_time,
+                    is_freed: false.into(),
                 });
             } else {
                 join_all(
@@ -274,6 +285,7 @@ impl LockManager {
                 .map(|client| Self::unlock_instance(client, &lock.resource, &lock.val)),
         )
         .await;
+        lock.is_freed.store(true, Ordering::Release);
     }
 
     /// Acquire the lock for the given resource and the requested TTL.
@@ -484,6 +496,7 @@ mod tests {
             resource: key,
             val,
             validity_time: 0,
+            is_freed: false.into(),
         };
 
         rl.unlock(&lock).await;
@@ -754,5 +767,38 @@ mod tests {
             }
         });
         let _ = j.await;
+    }
+    #[tokio::test]
+    async fn test_is_freed_flag() {
+        let (_containers, addresses) = create_clients();
+        let rl = LockManager::new(addresses.clone());
+
+        let lock1 = rl
+            .lock(b"resource_1", std::time::Duration::from_millis(1000))
+            .await
+            .unwrap();
+
+        let lock1 = Arc::new(lock1);
+        // Send the lock and entry through the channel
+        let (tx, mut rx) = tokio::sync::mpsc::channel(32);
+        tx.send(("some info", lock1.clone(), rl.clone()))
+            .await
+            .unwrap();
+
+        let j = tokio::spawn(async move {
+            // Retrieve from channel and use
+            if let Some((_entry, lock1, rl)) = rx.recv().await {
+                rl.unlock(&lock1).await;
+            }
+        });
+        let _ = j.await;
+        assert!(lock1.is_freed());
+
+        let lock2 = rl
+            .lock(b"resource_2", std::time::Duration::from_millis(1000))
+            .await
+            .unwrap();
+        rl.unlock(&lock2).await;
+        assert!(lock2.is_freed());
     }
 }

